elyra/lsp/client.py

```diff
@@ -0,0 +1,100 @@
+#
+# Copyright 2018-2021 Elyra Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import time
+import json
+import subprocess
+import threading
+
+
+class ReadPipe(threading.Thread):
+    def __init__(self, pipe):
+        threading.Thread.__init__(self)
+        self.pipe = pipe
+        self.done = False
+        self.res = None
+
+    def wait(self):
+        start = time.time()
+        while not self.done:
+            end = time.time()
+            if end > start + 3:
+                raise TimeoutError()
+            pass
+        return self.res
+
+    def run(self):
+        message_size = None
+        while True:
+            line = self.pipe.readline()
+            line = line.decode("utf-8")
+            line = line[:-len("\r\n")]
+            if line == "":
+                break
+
+            if line.startswith("Content-Length: "):
+                line = line[len("Content-Length: "):]
+                message_size = int(line)
+
+        jsonrpc_res = self.pipe.read(message_size).decode("utf-8")
+        self.res = json.loads(jsonrpc_res)
+        self.done = True
+
+
+class LSPClient:
```

Comment:
Not sure how to phrase this, is the front end e..g VPE going to be communicating with the LSP in real time or just a one-off validation on submission? e.g. triggering validation when something is dragged or changed on the canvas. If so do we need to implement more of the ls protocol in the client? like  `"method": "textDocument/didChange",`

Reply:
The LSP and the VPE use (this validation package)[https://github.com/bourdakos1/pipeline-editor/tree/create-node/packages/pipeline-services/src/validation] for validation logic. Right now this only uses the LSP for validation at submission time, while the VPE uses the validation package directly in real time. I don't know if we will ever have the VPE use the LSP directly since it has full access to the validation logic without having to make requests to the server